File: leetcode/cpp/Best Time to Buy and Sell Stock/Best Time to Buy and Sell Stock.cpp
class Solution {
public:
    //Dynamic Programming
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        if(n == 0)
            return 0;
        int max = 0;
        int lowest = prices[0];
        for(int i = 0; i < n; ++i){
            int tmp = prices[i] - lowest;
            if(tmp > max){
                max = tmp;
            }
            if(prices[i] < lowest){
                lowest = prices[i];
            }
        }
        return max;
    }
};

File: leetcode/python3/best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_val = 2**32
        max_val = 0
        for i in range(len(prices)):
            min_val = min(min_val, prices[i])
            max_val = max(max_val, prices[i]-min_val)
        return max_val

